### PR TITLE
Add functionality for postgresql database

### PR DESCRIPTION
**Pull request by tanweeralii**
Issue #13

### CODE REVIEW

File: src/database/inquirer.js [hunk 1]
```diff
@@ -71,6 +71,7 @@ let askConfig = async (jobName, key) => {
             let defaultPort;
             if (ans.dbType == 'MongoDB') defaultPort = '27017';
             else if (ans.dbType == 'MySQL') defaultPort = '3306';
+            else if(ans.dpType == 'PostgreSQL') defaultPort = '5432';
```
Comment (hariprasanths) [inline src/database/inquirer.js]: Typo - `dptype`

File: src/database/validator.js [hunk 2]
```diff
@@ -22,7 +22,7 @@ const validateInitConfig = async (config) => {
     if (!config[dbConfigKeys.dbType]) {
         throw new Error(`Invalid config: Missing required field - '${dbConfigKeys.dbType}'`);
     }
-    if (!['MongoDB', 'MySQL'].includes(config[dbConfigKeys.dbType])) {
+    if (!['MongoDB', 'MySQL', 'PostgreSQL'].includes(config[dbConfigKeys.dbType])) {
```
Comment (hariprasanths) [inline src/database/validator.js:25]: Update README under features to show that PostgreSQL is supported and examples to include PostgreSQL sample 
configuration.

**Review by hariprasanths — CHANGES_REQUESTED** [review]

**Comment by hariprasanths** [comment]
Could you please make the commit messages imperative?

Try to follow these guidelines when writing commit messages:
1. Separate subject from body with a blank line
2. Limit the subject line to 50 characters
3. Capitalize the subject line
4. Do not end the subject line with a period
5. Use the imperative mood in the subject line
6. Wrap the body at 72 characters
7. Use the body to explain what and why vs. how

Source: https://chris.beams.io/posts/git-commit/

**Comment by tanweeralii** [comment]
Ok sure

File: src/database/database.js [hunk 3]
```diff
@@ -2,6 +2,7 @@ const strings = require('./../utils/strings');
 const configstore = require('conf');
 const mongoDb = require('./mongoDb/mongoDb');
 const mysql = require('./mysql/mysql');
+const postgresql = require('./postgresql/postgrsql');
```
Comment (hariprasanths) [inline src/database/database.js]: Typo

File: src/database/postgresql/postgresql.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    });
+
+    const connRes = await awaitPostgresqlConnect(connection);
+    const disConnRes = await awaitPostgresqlEnd(connection);
+
+    return connRes;
+};
+
+let dump = async (dbConfig, key, backupPath) => {
+    const postgresqlDumpCmd = `PGPASSWORD=${dbConfig.dbAuthPwd} \
```
Comment (hariprasanths) [inline src/database/postgresql/postgresql.js]: Make this command compatible for powershell(windows) as well.

File: src/database/postgresql/postgresql.js [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        await files.decrypt(backupFilePath, key);
+        backupFilePath = `${backupFilePath}_unenc`;
+        decryptFileName = backupFilePath;
+    }
+    let isCompressed = isGzip(backupFilePath);
+    if (isCompressed) {
+        await files.decompressFile(backupFilePath);
+        backupFilePath = `${backupFilePath}.sql`;
+    }
+    const postgresqlRestoreCmd = `PGPASSWORD=${dbConfig.dbAuthPwd} \
```
Comment (hariprasanths) [inline src/database/postgresql/postgresql.js]: Make this command compatible for powershell(windows) as well.

**Review by hariprasanths — APPROVED** [review]

File: src/database/postgresql/postgresql.js (hunk 5 above)
Comment (tanweeralii) [reply src/database/postgresql/postgresql.js]: ok

File: src/database/database.js (hunk 3 above)
Comment (hariprasanths) [reply src/database/database.js]: Is this resolved?

Comment (hariprasanths) [reply src/database/database.js]: Do test the application before committing.

**Review by hariprasanths — APPROVED** [review]

File: src/database/database.js (hunk 3 above)
Comment (tanweeralii) [reply src/database/database.js]: ohhh I am doing it right now

File: src/database/postgresql/postgresql.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    const connRes = await awaitPostgresqlConnect(connection);
+    const disConnRes = await awaitPostgresqlEnd(connection);
+
+    return connRes;
+};
+
+let dump = async (dbConfig, key, backupPath) => {
+    const postgresqlDumpCmd = `pg_dump --dbname=postgresql://${dbConfig.dbAuthUser}:${dbConfig.dbAuthPwd}@${dbConfig.dbHost}:${dbConfig.dbPort}/${dbConfig.dbName} \
+    --compress=0..9 \
+    --format=c
```
Comment (hariprasanths) [inline src/database/postgresql/postgresql.js]: Escape the new line

Comment (hariprasanths) [reply src/database/postgresql/postgresql.js]: Always test the features before committing.

**Review by hariprasanths — CHANGES_REQUESTED** [review]

File: src/database/postgresql/postgresql.js (hunk 6 above)
Comment (tanweeralii) [reply src/database/postgresql/postgresql.js]: Yeah I did. I have checked this file and is not showing me this error.

Comment (tanweeralii) [reply src/database/postgresql/postgresql.js]: Ok. I fixed it but I will commit after testing the feature

Comment (hariprasanths) [reply src/database/postgresql/postgresql.js]: Okay, the backup command doesnt work for me (in bash shell) without the escaping the new line.
Try running the synchly commands to test out the features.
For example: Backup a db using the `synchly --run` command to take an instant backup and restore it using `synchly -R`.

Comment (tanweeralii) [reply src/database/postgresql/postgresql.js]: ok

Comment (tanweeralii) [reply src/database/postgresql/postgresql.js]: Hey, I tried doing backup and restore. I am getting error in restore but backup is successful.

Comment (tanweeralii) [reply src/database/postgresql/postgresql.js]: for every table present in database it is showing that the table already exists.

Comment (tanweeralii) [reply src/database/postgresql/postgresql.js]: Screenshot from 2020-11-22 15-52-55

**Comment by hariprasanths** [comment]
Ohh, try dropping the db before restoring.
I remember seeing support for the same in pg_restore command.

**Comment by tanweeralii** [comment]
Now both dump and restore functions are working properly.

File: package.json [hunk 7]
```diff
@@ -60,17 +60,18 @@
...
-        "mongoose": "^5.9.11",
+        "mongoose": "^5.10.15",
         "mysql": "^2.18.1",
-        "pg": "^8.5.1",
         "node-cron": "^2.0.3",
-        "nodemailer": "^6.4.6",
-        "omelette": "^0.4.12",
-        "ora": "^4.0.4",
-        "ssh2-sftp-client": "^5.1.2"
+        "node-mailer": "^0.1.1",
```
Comment (hariprasanths) [inline package.json]: Can you check this dependency. It does not look like the package we are using for nodemailer.

Comment (tanweeralii) [reply package.json]: sure.

File: package.json [hunk 8]
```diff
@@ -60,18 +60,17 @@
...
         "keytar": "^6.0.1",
-        "mongoose": "^5.10.15",
+        "mongoose": "^5.9.11",
         "mysql": "^2.18.1",
         "node-cron": "^2.0.3",
-        "node-mailer": "^0.1.1",
-        "nodemailer": "^6.4.16",
-        "omelette": "^0.4.15",
-        "ora": "^4.1.1",
+        "node-mailer": "^6.4.6",
```
Comment (hariprasanths) [inline package.json]: Kindly verify this package.
Seems like [this](https://www.npmjs.com/package/nodemailer) is the official package.

Comment (tanweeralii) [reply package.json]: ok sure

Comment (tanweeralii) [reply package.json]: done

File: package.json [hunk 9]
```diff
@@ -66,7 +66,7 @@
         "mongoose": "^5.9.11",
         "mysql": "^2.18.1",
         "node-cron": "^2.0.3",
-        "node-mailer": "^6.4.6",
+        "nodemailer": "^6.4.6",
```
Comment (hariprasanths) [inline package.json]: Can you please check the version of nodemailer. Seems like 6.4.17 is the latest version. [Nodemailer documentation](https://www.npmjs.com/package/nodemailer)

Comment (tanweeralii) [reply package.json]: sure

**Comment by tanweeralii** [comment]
I think postgresql part is completed. Now i will start working on s3 feature.

**Comment by hariprasanths** [comment]
Can you please remove the package-lock.json file from the changes.

**Pull request merged**